compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java

```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
     private ImmutableBiMap<ComponentDescriptor, String> disambiguateConflictingSimpleNames(
         Collection<ComponentDescriptor> components) {
       Map<String, ComponentDescriptor> generatedSimpleNames = new LinkedHashMap<>();
-      // The ending condition is when there is a unique simple name generated for every element
-      // in components. The sizes should be equivalent (with one generated name per component).
-      for (int levels = 0; generatedSimpleNames.size() != components.size(); levels++) {
+
+      // Let's see if we can get away with using simpleName() everywhere.
+      for (ComponentDescriptor component : components) {
+        Namer namer = componentNamers.get(component);
+        if (generatedSimpleNames.containsKey(namer.simpleName())) break;
+        generatedSimpleNames.put(namer.simpleName(), component);
+      }
+
+      if (generatedSimpleNames.size() != components.size()) {
+        // Simple approach didn't work out, let's use more complicated names.
+        // We keep them small to fix https://github.com/google/dagger/issues/421.
         generatedSimpleNames.clear();
+        UniqueNameSet nameSet = new UniqueNameSet();
         for (ComponentDescriptor component : components) {
-          List<String> pieces = componentQualifiedNamePieces.get(component);
-          String simpleName =
-              QUALIFIED_NAME_JOINER.join(
-                  pieces.subList(Math.max(0, pieces.size() - levels - 1), pieces.size()));
-          ComponentDescriptor conflict = generatedSimpleNames.put(simpleName, component);
-          if (conflict != null) {
-            // if the map previously contained an entry for the same simple name, stop early since
-            // 2+ subcomponent descriptors will have the same simple name
-            break;
-          }
+          Namer namer = componentNamers.get(component);
+          String simpleName = namer.simpleName();
+          String basePrefix = namer.uniquingPrefix();
+          generatedSimpleNames.put(format("%s_%s", nameSet.getUniqueName(basePrefix), simpleName),
+              component);
         }
       }
       return ImmutableBiMap.copyOf(generatedSimpleNames).inverse();
     }
 
-    private static ImmutableListMultimap<ComponentDescriptor, String> qualifiedNames(
+    private static ImmutableMap<ComponentDescriptor, Namer> qualifiedNames(
         Iterable<ComponentDescriptor> componentDescriptors) {
-      ImmutableListMultimap.Builder<ComponentDescriptor, String> builder =
-          ImmutableListMultimap.builder();
+      ImmutableMap.Builder<ComponentDescriptor, Namer> builder = ImmutableMap.builder();
       for (ComponentDescriptor component : componentDescriptors) {
-        Name qualifiedName = component.componentDefinitionType().getQualifiedName();
-        builder.putAll(component, QUALIFIED_NAME_SPLITTER.split(qualifiedName));
```

Comment:
Won't this actually result in longer names when simple names are used now that we're not splitting the qualified name?